### ISSUE TITLE
X10 slower when computing Hessian-vector product multiple times

### ISSUE THREAD

**Issue by zhangyilang** [pytorch/pytorch#76201]
### 🐛 Describe the bug

When computing the Hessian-vector product using `torch.autograd.grad` for multiple times, the subsequent computations would slow down by more than 10 times. Below is a simple example on a 4-layer CNN:
```python
from timeit import default_timer as timer
import torch
import torch.nn as nn


def conv3x3(in_channels, out_channels):
    return nn.Sequential(
        nn.Conv2d(in_channels, out_channels, 3, padding=1),
        nn.BatchNorm2d(32),
        nn.ReLU(),
        nn.MaxPool2d(2),
    )

class CNN(nn.Module):
    def __init__(self):
        super(CNN, self).__init__()
        self.network = nn.Sequential(
            conv3x3(3, 32),
            conv3x3(32, 32),
            conv3x3(32, 32),
            conv3x3(32, 32),
            nn.Flatten(),
            nn.Linear(800, 10)
        )

    def forward(self, x):
        out = self.network(x)

        return out


if __name__ == '__main__':
    device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
    model = CNN().to(device)
    params = tuple(model.parameters())
    inputs = torch.randn(80, 3, 84, 84).to(device)
    vec = torch.randn_like(nn.utils.parameters_to_vector(params))

    outputs = model(inputs)
    loss = (outputs**2).sum()
    grads = torch.autograd.grad(loss,
                                params,
                                create_graph=True,
                                retain_graph=True,
                                allow_unused=True)
    grad_vec = torch.cat([g.view(-1) for g in grads])

    for _ in range(10):
        start = timer()
        hvp = torch.autograd.grad(grad_vec,
                                  params,
                                  grad_outputs=vec,
                                  create_graph=True,
                                  retain_graph=True,
                                  allow_unused=True)
        end = timer()
        print(end - start)
        del hvp

```
By running the codes, I got the following results:
```
0.010107266018167138
0.023112661205232143
0.1425181080121547
0.1398529140278697
0.13755728211253881
0.14755830401554704
0.15265934192575514
0.1391770471818745
0.13788723316974938
0.14136399910785258
```
It is pretty weird that the first, second and subsequent computations take about 0.01s, 0.02s and 0.14s, respectively. I also checked that the maximum GPU memory occupied (through `torch.cuda.max_memory_allocated()`) is only 1839M, which is far smaller than my GPU capacity. Therefore, it seems that the degraded performances are not caused by the GPU memory.

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (Ubuntu 6.5.0-2ubuntu1~18.04) 6.5.0 20181026
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.9.7 (default, Sep 16 2021, 13:09:58)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-4.15.0-175-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: NVIDIA GeForce GTX TITAN X
Nvidia driver version: 470.103.01
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] torch==1.11.0
[pip3] torchaudio==0.11.0
[pip3] torchvision==0.12.0
[conda] cudatoolkit               11.3.1               h2bc3f7f_2

**Comment by vadimkantorov** [comment]
(just to be sure, please use torch.cuda.synchronize() calls to ensure that CUDA computation is finished when measuring time)

**Comment by zhangyilang** [comment]
After adding `torch.cuda.synchronize()`, I got the following results
```
0.08550256816670299
0.07527337200008333
0.07538111205212772
0.0752657309640199
0.07532073208130896
0.07529082102701068
0.07521857600659132
0.07531275204382837
0.07537108985707164
0.07537124585360289
```
which seems good. 

However, I also found that computing the gradients only takes about 0.017s. It seems the HVP computations still take more time compared to the gradients. Any idea to narrow this gap?

**Comment by wagenaartje** [comment]
I would be interested in a faster way to calculate Hessian vector products too, as @zhangyilang said it takes too much time compared to plain gradients.
